### PR TITLE
API 5314 Smart Configuration

### PR DESCRIPTION
**Pull request by that-guy-is-it**

### CODE REVIEW

File: smart-cards/src/test/java/gov/va/api/health/smartcards/WellKnownControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.Test;
+
+public class WellKnownControllerTest {
+  private WellKnown actual() {
+    return WellKnown.builder()
+        .tokenEndpoint("https://somevasite.va.gov/token")
+        .authorizationEndpoint("https://somevasite.va.gov/authorize")
+        .managementEndpoint("https://somevasite.va.gov/manage")
+        .revocationEndpoint("https://somevasite.va.gov/revoke")
```
Comment (apeirats-va) [inline smart-cards/src/test/java/gov/va/api/health/smartcards/WellKnownControllerTest.java:18]: Auth endpoints are already published by DQ, so it's likely unnecessary to re-publish them here.

File: smart-cards/src/main/resources/application.properties [hunk 2]
```diff
@@ -2,9 +2,16 @@ data-query.internal-url=unset
...
+metadata.security.revocation-endpoint=unset
 public-r4-base-path=unset
 public-url=unset
 server.port=8096
 spring.autoconfigure.exclude=gov.va.api.health.autoconfig.configuration.SecureRestTemplateConfig
 ssl.enable-client=false
 web-exception-key=unset
+well-known.capabilities=unset
+well-known.response=unset
+well-known.scopes-supported=unset
```
Comment (apeirats-va) [inline smart-cards/src/main/resources/application.properties:17]: Are these different per environment? If not, we can set them all here and not deal with DU.

File: make-configs.sh [hunk 3]
```diff
@@ -40,7 +40,14 @@ configValue smart-cards $PROFILE data-query.internal-url 'http://localhost:8777'
...
+configValue smart-cards $PROFILE metadata.security.token-endpoint 'http://fake.com/token'
+configValue smart-cards $PROFILE metadata.security.authorize-endpoint 'http://fake.com/authorize'
+configValue smart-cards $PROFILE metadata.security.management-endpoint 'http://fake.com/manage'
+configValue smart-cards $PROFILE metadata.security.revocation-endpoint 'http://fake.com/revoke'
 configValue smart-cards $PROFILE public-r4-base-path 'r4'
 configValue smart-cards $PROFILE public-url 'http://localhost:8096'
 configValue smart-cards $PROFILE web-exception-key '-sharktopus-v-pteracuda-'
+configValue smart-cards $PROFILE well-known.capabilities "context-standalone-patient, launch-ehr, permission-offline, permission-patient"
+configValue smart-cards $PROFILE well-known.response "ode, refresh_token"
+configValue smart-cards $PROFILE well-known.scopes-supported "patient/DiagnosticReport.read, patient/Patient.read, offline_access"
```
Comment (apeirats-va) [inline make-configs.sh:52]: Compare this to example on spec website:
https://smarthealth.cards/#via-fhir-health-cards-issue-operation

I think the only difference between SC and DQ is the `health-cards` capability and the scopes supported.

File: smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import gov.va.api.health.r4.api.information.WellKnown;
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(
+    value = {"/.well-known/smart-configuration"},
```
Comment (apeirats-va) [inline smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java:12]: Should this be under `/r4/.well-known/...`?

Comment (apeirats-va) [reply smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java:12]: Nevermind. Keep this at top level, unless there's any differences between r4 and dstu2 versions (there shouldn't be). The unifier exposes the public endpoint at the correct endpoint.

File: smart-cards/src/main/resources/application.properties [hunk 5]
```diff
@@ -2,9 +2,16 @@ data-query.internal-url=unset
...
+metadata.security.revocation-endpoint=unset
 public-r4-base-path=unset
 public-url=unset
 server.port=8096
 spring.autoconfigure.exclude=gov.va.api.health.autoconfig.configuration.SecureRestTemplateConfig
 ssl.enable-client=false
 web-exception-key=unset
+well-known.capabilities="health-cards, launch-standalone, context-standalone-patient, client-confidential-symmetric"
+well-known.response="ode, refresh_token"
+well-known.scopes-supported="launch, launch/patient, patient/*.*, offline_access"
```
Comment (apeirats-va) [inline smart-cards/src/main/resources/application.properties:17]: let's expand the wildcards:
```
patient/Patient.read patient/Immunization.read patient/Location.read
```

File: smart-cards/src/main/resources/application.properties [hunk 6]
```diff
@@ -2,9 +2,16 @@ data-query.internal-url=unset
...
+metadata.security.management-endpoint=unset
+metadata.security.revocation-endpoint=unset
 public-r4-base-path=unset
 public-url=unset
 server.port=8096
 spring.autoconfigure.exclude=gov.va.api.health.autoconfig.configuration.SecureRestTemplateConfig
 ssl.enable-client=false
 web-exception-key=unset
+well-known.capabilities="health-cards, launch-standalone, context-standalone-patient, client-confidential-symmetric"
+well-known.response="ode, refresh_token"
```
Comment (apeirats-va) [inline smart-cards/src/main/resources/application.properties:16]: Typo: `ode` -> `code`

File: smart-cards/src/main/resources/application.properties (hunk 2 above)
Comment (jblefkove-va) [reply smart-cards/src/main/resources/application.properties]: 👍 and I'll take it a step further: if these values are the same for all environments, skip app properties altogether and hardcode them directly in `WellKnownController`

File: smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    produces = {"application/json", "application/fhir+json", "application/json+fhir"})
+@AllArgsConstructor(onConstructor = @__({@Autowired}))
+public class WellKnownController {
+
+  private final WellKnownProperties wellKnownProperties;
```
Comment (apeirats-va) [inline smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java:17]: We can remove WellKnownProperties too, and have the controller build it all with hardcoded values.

File: smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(
+    value = {"/.well-known/smart-configuration"},
+    produces = {"application/json", "application/fhir+json", "application/json+fhir"})
```
Comment (jblefkove-va) [inline smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java:13]: remove `json+fhir`

File: smart-cards/src/test/java/gov/va/api/health/smartcards/WellKnownControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  private WellKnown actual() {
+    return WellKnown.builder()
+        .capabilities(
+            asList(
+                "context-standalone-patient",
+                "launch-standalone",
+                "permission-offline",
+                "permission-patient"))
+        .responseTypeSupported(asList("code", "refresh-token"))
+        .scopesSupported(asList("patient/$HealthWallet.issueVc"))
```
Comment (jblefkove-va) [inline smart-cards/src/test/java/gov/va/api/health/smartcards/WellKnownControllerTest.java:22]: update this to `$health-cards-issue`

Comment (jblefkove-va) [reply smart-cards/src/test/java/gov/va/api/health/smartcards/WellKnownControllerTest.java:22]: Oh this is only in the test... remove.

File: smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+          "launch/patient",
+          "patient/Patient.read",
+          "patient/Immunization.read",
+          "patient/Location.read",
+          "offline_access");
+
+  @GetMapping
+  WellKnown read() {
+    return WellKnown.builder()
+        .capabilities(capabilities)
```
Comment (jblefkove-va) [inline smart-cards/src/main/java/gov/va/api/health/smartcards/WellKnownController.java:43]: We can simplify further, inline `capabilities`, `responseTypeSupported`, etc. directly in this method.

**Review by jblefkove-va — APPROVED** [review]

**Pull request merged**